aggregator/src/aggregator/key_rotator.rs

```diff
@@ -0,0 +1,673 @@
+                                .await?;
+                        }
+                    }
+                }
+            }
+
+            // Promote any pending keypairs that are ready. We don't sweep them in the same loop
+            // as active keypairs, because it's possible for there to be a pending keypair
+            // without an active one.
+            for pending_keypair in pending_keypairs {
+                if Self::duration_since(tx, pending_keypair.updated_at()) > config.pending_duration
+                {
+                    info!(id = ?pending_keypair.id(), "pending key is ready, moving it to active");
+                    tx.set_global_hpke_keypair_state(pending_keypair.id(), &HpkeKeyState::Active)
+                        .await?;
+                }
+            }
+        }
+
+        // Sweep expired keypairs. Any keypairs that are old enough are deleted.
+        for keypair in &expired_keypairs {
+            if Self::duration_since(tx, keypair.updated_at()) > config.expired_duration {
+                info!(id = ?keypair.id(), "deleting expired keypair");
+                tx.delete_global_hpke_keypair(keypair.id()).await?;
+            }
+        }
+
+        Ok(())
+    }
+
+    fn duration_since(tx: &Transaction<C>, time: &Time) -> Duration {
+        // Use saturating difference to account for time skew between key rotator runners. Since
+        // key rotators are synchronized by an exclusive lock on the table, it's possible that
+        // time skew between concurrently running replicas result in underflows.
+        tx.clock().now().saturating_difference(time)
+    }
+
+    fn partition_keypairs<'a>(
+        keypairs: &'a [GlobalHpkeKeypair],
+        ciphersuite: &HpkeCiphersuite,
+    ) -> (
+        Vec<&'a GlobalHpkeKeypair>,
+        Vec<&'a GlobalHpkeKeypair>,
+        Vec<&'a GlobalHpkeKeypair>,
+    ) {
+        let (mut pending, mut active, mut expired) = (Vec::new(), Vec::new(), Vec::new());
+        keypairs
+            .iter()
+            .filter(|keypair| &keypair.ciphersuite() == ciphersuite)
+            .for_each(|keypair| match keypair.state() {
+                HpkeKeyState::Pending => pending.push(keypair),
+                HpkeKeyState::Active => active.push(keypair),
+                HpkeKeyState::Expired => expired.push(keypair),
+            });
+        (pending, active, expired)
+    }
+
+    fn available_ids(current_keypairs: &[GlobalHpkeKeypair]) -> impl Iterator<Item = HpkeConfigId> {
+        let ids: Vec<_> = current_keypairs
+            .iter()
+            .map(|keypair| u8::from(*keypair.id()))
+            .collect();
+
+        // Try to cycle through the entire u8 space before going back to zero. This allows us a
+        // bigger window to quicky reject faulty old clients with an outdated HPKE config error,
+        // rather than attempting to decrypt them with the wrong key.
```

Comment:
nit
```suggestion
        // bigger window to quicky reject reports from faulty old clients with an outdated HPKE
        // config error, rather than attempting to decrypt them with the wrong key.
```